### PR TITLE
[ring] Ring Binding initial contribution

### LINKED ISSUE #8904: Openhab 3 compatible Ring Binding - Please add binding to source code
Hi Can we please request for the source code of the Ring Binding to be added to OH3.

Can the current binding be fixed with a PR. 

Code below by 

https://github.com/zolakk/openhab2-addons-ring
 
Code provide by zolakk [https://github.com/zolakk/openhab2-addons-ring/commits?author=zolakk](url)

## Your Environment

Openhab 3
Windows 10
Java 11

### PR DESCRIPTION
**Pull request by morph166955**
This is the initial contribution of the Ring binding.  There was a version for OH2.5 that this is derived from.  Code was written circa 2020 by Wim Vissers and all credit is given to him for the original authoring.  It appears that it was then adapted by 
Sriram Balakrishnan for OH3 in December of 2021.  My fork was created from the commit found here... https://github.com/openhab/openhab-addons/commit/56b75c26db597158191e407b5119a8860bc3dc8f

All indications seem to be that nothing has been done to the code since then.  I've volunteered to pick this up as the maintainer and pull it into OH4 as there is a community (myself included) who use Ring.  Code has been adapted and several issues resolved along the way. 

JAR is available through the marketplace at: https://community.openhab.org/t/ring-binding-3-2-0-4-0-0/145245

Fixes #8904

### CODE REVIEW

**Comment by morph166955** [comment]
Exception randomly throwing:

```
2023-03-13 16:34:53.255 [ERROR] [hab.binding.ring.internal.RestClient] - RestApi error in getRequest!
java.net.SocketTimeoutException: Connect timed out
        at sun.nio.ch.NioSocketImpl.timedFinishConnect(NioSocketImpl.java:546) ~[?:?]
        at sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:597) ~[?:?]
        at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:327) ~[?:?]
        at java.net.Socket.connect(Socket.java:633) ~[?:?]
        at sun.security.ssl.SSLSocketImpl.connect(SSLSocketImpl.java:304) ~[?:?]
        at sun.net.NetworkClient.doConnect(NetworkClient.java:178) ~[?:?]
        at sun.net.www.http.HttpClient.openServer(HttpClient.java:531) ~[?:?]
        at sun.net.www.http.HttpClient.openServer(HttpClient.java:636) ~[?:?]
        at sun.net.www.protocol.https.HttpsClient.<init>(HttpsClient.java:266) ~[?:?]
        at sun.net.www.protocol.https.HttpsClient.New(HttpsClient.java:380) ~[?:?]
        at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.getNewHttpClient(AbstractDelegateHttpsURLConnection.java:193) ~[?:?]
        at sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1242) ~[?:?]
        at sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1128) ~[?:?]
        at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:179) ~[?:?]
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1665) ~[?:?]
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1589) ~[?:?]
        at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:529) ~[?:?]
        at sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:308) ~[?:?]
        at org.openhab.binding.ring.internal.RestClient.getRequest(RestClient.java:219) ~[?:?]
        at org.openhab.binding.ring.internal.RestClient.getHistory(RestClient.java:617) ~[?:?]
        at org.openhab.binding.ring.handler.AccountHandler.minuteTick(AccountHandler.java:336) ~[?:?]
        at org.openhab.binding.ring.handler.AbstractRingHandler$1.run(AbstractRingHandler.java:84) ~[?:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) ~[?:?]
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305) ~[?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
        at java.lang.Thread.run(Thread.java:833) ~[?:?]
```

**Comment by morph166955** [comment]
Exception randomly throwing:

```
2023-03-14 07:43:06.556 [ERROR] [hab.binding.ring.internal.RestClient] - RestApi error in getRequest!
java.net.UnknownHostException: api.ring.com
        at sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:567) ~[?:?]
        at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:327) ~[?:?]
        at java.net.Socket.connect(Socket.java:633) ~[?:?]
        at sun.security.ssl.SSLSocketImpl.connect(SSLSocketImpl.java:304) ~[?:?]
        at sun.net.NetworkClient.doConnect(NetworkClient.java:178) ~[?:?]
        at sun.net.www.http.HttpClient.openServer(HttpClient.java:531) ~[?:?]
        at sun.net.www.http.HttpClient.openServer(HttpClient.java:636) ~[?:?]
        at sun.net.www.protocol.https.HttpsClient.<init>(HttpsClient.java:266) ~[?:?]
        at sun.net.www.protocol.https.HttpsClient.New(HttpsClient.java:380) ~[?:?]
        at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.getNewHttpClient(AbstractDelegateHttpsURLConnection.java:193) ~[?:?]
        at sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1242) ~[?:?]
        at sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1128) ~[?:?]
        at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:179) ~[?:?]
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1665) ~[?:?]
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1589) ~[?:?]
        at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:529) ~[?:?]
        at sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:308) ~[?:?]
        at org.openhab.binding.ring.internal.RestClient.getRequest(RestClient.java:219) ~[?:?]
        at org.openhab.binding.ring.internal.RestClient.getRingDevices(RestClient.java:600) ~[?:?]
        at org.openhab.binding.ring.handler.AccountHandler.refreshRegistry(AccountHandler.java:278) ~[?:?]
        at org.openhab.binding.ring.handler.AccountHandler.minuteTick(AccountHandler.java:288) ~[?:?]
        at org.openhab.binding.ring.handler.AbstractRingHandler$1.run(AbstractRingHandler.java:84) ~[?:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) ~[?:?]
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305) ~[?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
        at java.lang.Thread.run(Thread.java:833) ~[?:?]

```

File: bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/handler/AccountHandler.java
Comment (clinique) [inline bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/handler/AccountHandler.java:205]: Maybe it would be worth declaring a configuration class and then `AccountConfiguration config = getThing().getConfiguration().as(AccountConfiguration.class);` - this would save you from below lines

File: bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/handler/AccountHandler.java
Comment (clinique) [inline bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/handler/AccountHandler.java:226]: Tbc confirmed : no need to `remove` if you `put`.

File: bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/handler/AccountHandler.java
Comment (clinique) [inline bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/handler/AccountHandler.java:234]: idem

File: bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/handler/AccountHandler.java
Comment (clinique) [inline bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/handler/AccountHandler.java:285]: remove comment

File: bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/handler/AccountHandler.java
Comment (clinique) [inline bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/handler/AccountHandler.java:294]: Worth defining a dedicated configuration class here also.

File: bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/internal/RingHandlerFactory.java
Comment (clinique) [inline bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/internal/RingHandlerFactory.java:78]: `SUPPORTED_THING_TYPES_UIDS = Set.of(THING_TYPE_ACCOUNT, THING_TYPE_DOORBELL.....);`

**Review by clinique — CHANGES_REQUESTED** [review]
Starting a code review

File: bundles/org.openhab.binding.ring/src/main/resources/OH-INF/binding/binding.xml
Comment (clinique) [inline bundles/org.openhab.binding.ring/src/main/resources/OH-INF/binding/binding.xml:1]: This file can be removed

File: bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/internal/discovery/RingDiscoveryService.java
Comment (clinique) [inline bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/internal/discovery/RingDiscoveryService.java:49]: See above comment regarding usage of `Set.of(....)`

File: bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/internal/data/RingDevices.java
Comment (clinique) [inline bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/internal/data/RingDevices.java:170]: replace by return `List.of(doorbells,stickupcams);`

**Review by clinique — CHANGES_REQUESTED** [review]

**Comment by morph166955** [comment]
@clinique thank you for the review!  I'll hit them next week when I get home from the beach. Any chance you have some time to do a pass over the AndroidTV binding so I can hit them both at the same time next week?

**Comment by clinique** [comment]
> @clinique thank you for the review!  I'll hit them next week when I get home from the beach. Any chance you have some time to do a pass over the AndroidTV binding so I can hit them both at the same time next week?

I'll try to give it a look. Have nice rest and 🏖️ time

**Comment by morph166955** [comment]
Thank you and greatly appreciated in advance!!!!

File: bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/internal/data/RingDevices.java
Comment (morph166955) [reply bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/internal/data/RingDevices.java:170]: [ERROR] /home/openhab/ring/openhab-addons/bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/internal/data/RingDevices.java:[167,35] Type mismatch: cannot convert from java.util.List<java.util.List<? extends org.openhab.binding.ring.internal.data.AbstractRingDevice>> to java.util.List<org.openhab.binding.ring.internal.data.RingDevice>

File: bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/handler/AbstractRingHandler.java
Comment (clinique) [reply bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/handler/AbstractRingHandler.java]: No typo, just could be initialized at declaration

File: bundles/org.openhab.binding.ring/src/main/resources/OH-INF/binding/binding.xml
Comment (morph166955) [reply bundles/org.openhab.binding.ring/src/main/resources/OH-INF/binding/binding.xml:1]: This is only in to have backwards compatibility with 3.4 for the marketplace.  Same as a few of the lines in pom.xml.  I will pull these out before merge.  This is mostly so I don't have to maintain multiple branches while testing as the bulk of the users are still 3.4.

File: bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/handler/AbstractRingHandler.java
Comment (morph166955) [reply bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/handler/AbstractRingHandler.java]: LOL I thought you were saying idem as in a typo of the word item...  I need more coffee.  I pulled them into declaration.

File: bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/handler/AccountHandler.java
Comment (morph166955) [reply bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/handler/AccountHandler.java:159]: I think I'm going to hold on this as to not creating a breaking change right now for the 3.4 folks since I am building jars for them.

**Comment by morph166955** [comment]
@clinique I hit most of the fast low hanging fruit changes.  The rest need a little more time than I have right now, I'll hit those later today or this week.  I'm going to defer the changes that break on 3.4 just to maintain compatibility for the folks testing.

File: bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/RingBindingConstants.java
Comment (morph166955) [reply bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/RingBindingConstants.java:39]: I went through the code and from what I can tell the original author was using it this way because of the way they structured the channel groups.  For example, battery is...

```
                        <channel-group id="status" typeId="deviceStatus"/>

        <channel-group-type id="deviceStatus">
                <label>Device Status</label>
                <description>Device Status Information</description>
                <channels>
                        <channel id="battery" typeId="battery"/>
                </channels>
        </channel-group-type>
```

And then updated as...

```
            ChannelUID channelUID = new ChannelUID(thing.getUID(), CHANNEL_STATUS_BATTERY);
            updateState(channelUID, new DecimalType(device.getBattery()));
```

Would you not need it to be status#battery in that case?

**Comment by morph166955** [comment]
@clinique I'm at an impasse after looking at the comments about a dedicated configuration class.  It looks like the original author was using the thing configuration to store some persistent items that the binding needs to periodically update (like a refresh token).  I have two issues with this.  

1) This will cause issues for folks who still use text based thing configurations (like me) as we will loose persistence.

2) I now get these errors at compile time because I'm now trying to pass my AccountConfiguration.class 

```
[ERROR] /home/openhab/ring/openhab-addons/bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/handler/AccountHandler.java:[223,17] The method updateConfiguration(org.openhab.core.config.core.Configuration) in the type org.openhab.core.thing.binding.BaseThingHandler is not applicable for the arguments (org.openhab.binding.ring.handler.AccountConfiguration)
[ERROR] /home/openhab/ring/openhab-addons/bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/handler/AccountHandler.java:[230,13] The method updateConfiguration(org.openhab.core.config.core.Configuration) in the type org.openhab.core.thing.binding.BaseThingHandler is not applicable for the arguments (org.openhab.binding.ring.handler.AccountConfiguration)
[ERROR] /home/openhab/ring/openhab-addons/bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/handler/AccountHandler.java:[240,13] The method updateConfiguration(org.openhab.core.config.core.Configuration) in the type org.openhab.core.thing.binding.BaseThingHandler is not applicable for the arguments (org.openhab.binding.ring.handler.AccountConfiguration)

``` 

I've not run into this yet on other bindings that I work on.  Is there a recommended way to fix this?

**Comment by clinique** [comment]
You could store your refresh token in the oh user data folder. Take a look at on of the last PR I pushed for Netatmo regarding auth evolution.

**Comment by morph166955** [comment]
For the time being, if I were to do something like this:

            config.refreshToken = userProfile.getRefreshToken();
            getThing().getConfiguration().put("refreshToken", config.refreshToken);
            updateConfiguration(getThing().getConfiguration());

Would that sufficiently update the token or do I need to store the Configuration and pass the updated one to updateConfiguration?

**Comment by morph166955** [comment]
Ok, I think I found a nice middle ground with the configuration.  It's compiling cleanly, I haven't done a regression test on it yet.  Let me know what you think of the updates.

**Comment by morph166955** [comment]
Found exception during debuging:

2023-04-03 17:10:34.074 [DEBUG] [ing.ring.handler.AbstractRingHandler] - Exception occurred during execution of startAutomaticRefresh(): Cannot invoke "String.length()" because "s" is null
java.lang.NullPointerException: Cannot invoke "String.length()" because "s" is null
        at java.io.StringReader.<init>(StringReader.java:51) ~[?:?]
        at org.json.simple.parser.JSONParser.parse(JSONParser.java:79) ~[?:?]
        at org.json.simple.parser.JSONParser.parse(JSONParser.java:75) ~[?:?]
        at org.openhab.binding.ring.internal.RestClient.getRingDevices(RestClient.java:623) ~[?:?]
        at org.openhab.binding.ring.handler.AccountHandler.refreshRegistry(AccountHandler.java:278) ~[?:?]
        at org.openhab.binding.ring.handler.AccountHandler.minuteTick(AccountHandler.java:287) ~[?:?]
        at org.openhab.binding.ring.handler.AbstractRingHandler$1.run(AbstractRingHandler.java:81) ~[?:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) ~[?:?]
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305) ~[?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
        at java.lang.Thread.run(Thread.java:833) ~[?:?]

**Comment by morph166955** [comment]
There are a few small things I'm waiting to hear testing results on from the marketplace, but this is more or less ready for a review at this time.

**Comment by morph166955** [comment]
> There are a few small things I'm waiting to hear testing results on from the marketplace, but this is more or less ready for a review at this time.

Testing looks successful minus one small glitch that hasn't been reproducible for diagnosis. Users report stable connection with expected data.

I have received a feature request to include 2 other device types I will likely try to add if there is time. If not I will submit a follow on PR.

File: bundles/org.openhab.binding.ring/pom.xml
Comment (wborn) [inline bundles/org.openhab.binding.ring/pom.xml:10]:
```suggestion
    <version>4.1.0-SNAPSHOT</version>
```

**Comment by morph166955** [comment]
This is updated for 4.1.0

**Comment by morph166955** [comment]
@openhab/add-ons-maintainers Requesting a review on this binding.  It's been floating around for a while on the forum, I've taken ownership of it and am looking to get it merged in.  Binding has been stable used in the community for a while now.  Thank you!

File: bundles/org.openhab.binding.ring/pom.xml
Comment (wborn) [inline bundles/org.openhab.binding.ring/pom.xml:10]:
```suggestion
    <version>4.2.0-SNAPSHOT</version>
```

**Review by lsiepel — CHANGES_REQUESTED** [review]
Thanks for picking this up! 
First pass, mainly looked at the documentation (readme, i18n and xml) When addressed i will look at the code itself. I'm no maintainer, but we might allready get this into shape before a final review. 
As this binding was probably created long time ago with older templates, you might want to at the current readme template as inspiration:
https://github.com/openhab/openhab-core/blob/main/tools/archetype/binding/src/main/resources/archetype-resources/README.md

File: bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/handler/AccountHandler.java
Comment (lsiepel) [reply bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/handler/AccountHandler.java:187]: As 4.0 and 4.1 have seen daylight and both are java17 only, you might want to make this change before merging this.

File: bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/handler/AccountHandler.java
Comment (lsiepel) [reply bundles/org.openhab.binding.ring/src/main/java/org/openhab/binding/ring/handler/AccountHandler.java:139]: This one is still open.

**Comment by morph166955** [comment]
@lsiepel Thank you for the first pass review!  I haven't forgotten about it, just trying to knock a few other things off first.  I hope to hit these some time in the next week.

**Comment by lsiepel** [comment]
> @lsiepel Thank you for the first pass review! I haven't forgotten about it, just trying to knock a few other things off first. I hope to hit these some time in the next week.

Ping when you are ready to proceed. 

Somehow i can't seem to checkout this PR to vscode, (Couldn't find branch 'ring' in remote 'github-desktop-morph166955'. A common reason for this is that the PR author has deleted their branch or their forked repository.)

Just some more comments.
- Please provide i18n placdeholders for Thing updateStatus details.
- Be sure the thing uid (discovery) uses the binding constants instead of hardcoded strings. 
- Check SAT for warnings

**Comment by morph166955** [comment]
That's odd. I can see it no issues at https://github.com/morph166955/openhab-addons/tree/ring

**Comment by lsiepel** [comment]
> That's odd. I can see it no issues at https://github.com/morph166955/openhab-addons/tree/ring

Both github desktop and vscode refuse to checkout from this PR, (error above), but straight from your fork is fine. Really weird, but i'll manage to complete the review :-)

**Comment by morph166955** [comment]
Very odd. Thank you for the review!  I'm hoping to knock out from here soon.

**Comment by lsiepel** [comment]
> how i can't seem to checkout this PR to vscode, (Couldn't find branch 'ring' in remote 'github-desktop-morph166955'. A common reason for this is that the PR author has deleted their branch or their forked repository.)
> 
> Just some m

Can you proceed with this PR?

**Comment by morph166955** [comment]
I'm hoping to have some time soon to whack through this.  Been very busy lately and haven't had the time to sit down to do any major development efforts.

File: bundles/org.openhab.binding.ring/pom.xml
Comment (wborn) [inline bundles/org.openhab.binding.ring/pom.xml:10]:
```suggestion
    <version>4.3.0-SNAPSHOT</version>
```

**Comment by lsiepel** [comment]
This binding seems to be a nice addition, so i hope it will get finsihed at some point in time. That being said as this binding has been here for a while without progress, it might get closed in the near future. Obviously the code is not gone and work can continue outside of this PR/repository. Once the work is done, i'm happy to review any newly created PR.

**Comment by morph166955** [comment]
I'm replacing my ring devices shortly with a different vendor so I won't be able to support this long term.  If someone wants to take this over they are welcome to.  I'll leave this here in the event someone wants it as a starting off point.

**Pull request closed without merging**
